resources/type-system/flow-analysis.md

```diff
@@ -0,0 +1,713 @@
+    - if `captured` is true then:
+      - `VM = VariableModel(declared, promoted, tested, true, false, captured)`.
+    - otherwise if `x` is promotable via initialization given `VM` then
+      - `VM = VariableModel(declared, [T], tested, true, false, captured)`.
+    - otherwise if `x` is promotable via assignment of `E` given `VM`
+      - `VM = VariableModel(declared, T::promoted, tested, true, false, captured)`.
+    - otherwise if `x` is demotable via assignment of `E` given `VM`
+      - `VM = VariableModel(declared, demoted, tested, true, false, captured)`.
+      - where `previous` is the suffix of `promoted` starting with the first type
+        `S` such that `T <: S`, and:
+        - if `S`is nullable and if `T <: Q` where `Q` is **NonNull(`S`)** then
+          `demoted` is `Q::previous`
+        - otherwise `demoted` is `previous`
+
+- `promote(E, T, M)` where `E` is an expression, `T` is a type which it may be
+  promoted to, and `M = FlowModel(r, VI)` is the flow model in which to promote,
+  is defined as follows:
+  - If `E` is not a promotion target, then `M`
+  - If `E` is a promotion target `x`, then
+    - Let `VM = VariableModel(declared, promoted, tested, assigned, unassigned,
+      captured)` be the variable model for `x` in `VI`
+    - If `x` is not promotable via type test to `T` given `VM`, then return `M`
+    - Else
+      - Let `S` be the current type of `x` in `VM`
+      - If `T <: S` then let `T1` = `T`
+      - Else if `S` is `X extends R` then let `T1` = `X & T`
+      - Else If `S` is `X & R` then let `T1` = `X & T`
+      - Else `x` is not promotable (shouldn't happen since we checked above)
+      - Let `VM2 = VariableModel(declared, T1::promoted, T1::tested, assigned,
+      unassigned, captured)`
+      - Return `FlowModel(r, VI[x -> VM2])`
+- `promoteToNonNull(E, M)` where `E` is an expression and `M` is a flow model is
+  defined to be `promote(E, T, M)` where `T0` is the type of `E`, and `T` is
+  **NonNull(`T0`)**.
+- `factor(T, S)` where `T` and `S` are types defines the "remainder" of `T` when
+  `S` has been removed from consideration by an instance check.  It is defined
+  as follows:
+  - If `T <: S` then `Never`
+  - Else if `T` is `R?` and `Null <: S` then `factor(R, S)`
+  - Else if `T` is `R?` then `factor(R, S)?`
+  - Else if `T` is `R*` and `Null <: S` then `factor(R, S)`
+  - Else if `T` is `R*` then `factor(R, S)*`
+  - Else if `T` is `FutureOr<R>` and `Future<R> <: S` then `factor(R, S)`
+  - Else if `T` is `FutureOr<R>` and `R <: S` then `factor(Future<R>, S)`
+
+
+Questions:
+ - The interaction between assignment based **promotion** and downwards inference is
+   probably managable.  I think doing downwards inference using the current
+   type, and then promoting the variable afterwards is fine for all reasonable
+   cases.
+ - The interaction between assignment based **demotion** and downwards inference is
```

Comment:
I have very little confidence that we're currently doing this correctly.  Do you think you could help come up with some test cases?

Reply:
Yes.